Hw23

Mykhaylo12 opened the pull request.

Mykhaylo12 committed:
hw23+24

Mykhaylo12 committed:
hw23+24

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/BucketDaoJdbcImpl.java:
```diff
@@ -0,0 +1,227 @@
+
+@Dao
+public class BucketDaoJdbcImpl extends AbstractDao<Bucket> implements BucketDao {
+    private static Logger LOGGER = Logger.getLogger(ItemDaoJdbcImpl.class);
```
```suggestion
    private static Logger LOGGER = Logger.getLogger(BucketDaoJdbcImpl.class);
```

Mykhaylo12 replied:
ок

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/BucketDaoJdbcImpl.java:
```diff
@@ -0,0 +1,227 @@
+            List<Item> items = new ArrayList<>();
+            Bucket bucket = new Bucket();
+            bucket.setBucketId(basketId);
+            bucket.setUserId(userId);
```
нам потрібно тут ці дві строки коду?

Mykhaylo12 replied:
видалив

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/BucketDaoJdbcImpl.java:
```diff
@@ -0,0 +1,227 @@
+    @Override
+    public Bucket update(Bucket bucket) {
+        List<Item> items = bucket.getItems();
+        deleteOldBucket(bucket.getBucketId());
```
не правильно видаляти а потім додавати бакет

Mykhaylo12 replied:
з айтемами  в ордері Богдан казав, що так може бути. Тут аналогічно

Sofasmile replied:
він напевно мав на увазі видаляти записи з суміжної таблиці, а потім додавати, а не повністю видаляти бакет)

Mykhaylo12 replied:
ok

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/BucketDaoJdbcImpl.java:
```diff
@@ -0,0 +1,227 @@
+            int rs = preparedStatement.executeUpdate();
+            if (rs > 0) {
+                return true;
+            } else {
```
чи можемо обійтись без блока else?

Mykhaylo12 replied:
ok

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/BucketDaoJdbcImpl.java:
```diff
@@ -0,0 +1,227 @@
+
+    @Override
+    public boolean deleteById(Long bucketId) {
+        long userId = bucketId;
```
????

Mykhaylo12 replied:
спростив

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/BucketDaoJdbcImpl.java:
```diff
@@ -0,0 +1,227 @@
+        }
+    }
+
+    private List<Item> getAllItemsByUserId(Long userId) {
```
нам потірбен цей метод?

Mykhaylo12 replied:
ні, але Богдан його почав робити на відео. Тому і я зробив. Можливо для майбутніх завдань?

Mykhaylo12 replied:
можу видалити якщо треба

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/BucketDaoJdbcImpl.java:
```diff
@@ -0,0 +1,227 @@
+
+    @Override
+    public Bucket create(Bucket bucket) {
+        bucket.setBucketId(bucket.getUserId());
```
чому ми в якості setBucketId використовуємо UserId?

Mykhaylo12 replied:
бо в юзера одночасно може бути тільки один бакет. Тому можна використовувати userId as bucketId. А ордерів може бути багато.

Sofasmile replied:
ні Міш, це погана практика, давай не будемо так робити. В нас можуть не співпадати ід юзера і бакета і потім можуть виникнути проблеми

Mykhaylo12 replied:
ok

Mykhaylo12 committed:
1)виправив рядок 21 BucketDaoJdbcImpl.class
2)BucketDaoJdbcImpl.java видалив дублюючий код bucket.setBucketId(basketId);bucket.setUserId(userId);
3)BucketDaoJdbcImpl.java лінія 138.Спростив блок if-else
4)забрав логери з DaoJdbcImpl рівня
5)поставив логери на рівні контроллерів
6)Зробив кастумний ексершин і DataProcessingException
7)Зробив error.jsp
8)Змінив таблицю ролів. Зробив по аналогії з ордерс

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/BucketDaoJdbcImpl.java:
```diff
@@ -112,80 +107,66 @@ public Bucket update(Bucket bucket) {
         String query = "DELETE FROM internet_shop.buckets WHERE user_id=?;";
         try (PreparedStatement preparedStatement = connection.prepareStatement(query)) {
             preparedStatement.setLong(1, userId);
             int rs = preparedStatement.executeUpdate();
```
execute() повертає true/false
можливо краще використовувати його?

Mykhaylo12 replied:
No. Because the execute method returns a boolean type value where true indicates that the ResultSet s returned which can later be extracted and false indicates that the integer or void value is returned. After delete I dont't  have resultSet

Mykhaylo12 committed:
1)виправив рядок 21 BucketDaoJdbcImpl.class
2)BucketDaoJdbcImpl.java видалив дублюючий код bucket.setBucketId(basketId);bucket.setUserId(userId);
3)BucketDaoJdbcImpl.java лінія 138.Спростив блок if-else
4)забрав логери з DaoJdbcImpl рівня
5)поставив логери на рівні контроллерів
6)Зробив кастумний ексершин і DataProcessingException i LoginExistExeption
7)Зробив error.jsp
8)Змінив таблицю ролів. Зробив по аналогії з ордерс
9)змінив таблицю бакетс на бакетс і бакетс_ітемс

Mykhaylo12 committed:
renamed LoginExistExeption LoginExistExCeption

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/BucketDaoJdbcImpl.java:
```diff
@@ -24,35 +24,72 @@ public BucketDaoJdbcImpl(Connection connection) {
+    }
+
+    private Optional<Bucket> checkIfExistBucket(Bucket bucket) throws DataProcessingException {
+        String query = "SELECT * FROM internet_shop.buckets WHERE user_id=?;";
```
а де ти встановлюєш значення user_id?

Mykhaylo12 replied:
після того як я перероби бакет, цей метод став не потрібним. Я його видалю при клінапі

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/BucketDaoJdbcImpl.java:
```diff
@@ -24,35 +24,72 @@ public BucketDaoJdbcImpl(Connection connection) {
+                    bucket.setItems(getAllItemsByUserId(bucket.getUserId()));
+                }
+                return Optional.of(bucket);
+            } else {
```
нам потрібен блок else тут?

Mykhaylo12 replied:
я цей метод видалю. Він вже не актуальний

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/BucketDaoJdbcImpl.java:
```diff
@@ -68,21 +105,21 @@ public Bucket create(Bucket bucket) throws DataProcessingException {
 
     @Override
     public Optional<Bucket> getByUser(User user) throws DataProcessingException {
```
методи get() і getByUser() дуже подібні. Можливо ми винесемо спілбну логіку в окремий метод?
а в цих методах лишиться щось подібне:
```java
    @Override
    public Optional<Bucket> getByUser(User user) throws DataProcessingException {
        String query = "SELECT buckets.bucket_id, buckets.user_id, buckets_items.item_id,"
                + " items.name, items.price FROM internet_shop.buckets_items INNER JOIN"
                + " internet_shop.buckets ON buckets_items.bucket_id=buckets.bucket_id INNER JOIN"
                + " internet_shop.items ON items.item_id=buckets_items.item_id"
                + " WHERE buckets.user_id=?;";
        return getUserByParameter(query, user.getId());
    }
```

Mykhaylo12 replied:
я спробую

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/UserDaoJdbcImpl.java:
```diff
@@ -47,6 +48,39 @@ public User create(User user) throws DataProcessingException {
         }
     }
 
+    public void checkUserLoginForRegistration(String login) throws LoginExistException {
```
ми можемо обійтись без цього методу)

Mykhaylo12 replied:
ці два методи , це порада Богдана

Sofasmile commented on src/main/java/mate/academy/internetshop/dao/jdbc/UserDaoJdbcImpl.java:
```diff
@@ -47,6 +48,39 @@ public User create(User user) throws DataProcessingException {
+    }
+
+    @Override
+    public void checkUserLoginForLogin(String login) throws DataProcessingException {
```
і цей також тут зайвий)

Mykhaylo12 replied:
ці два методи , це порада Богдана
